### PR TITLE
pkgconfig: Also pull InternalDependency.libraries

### PR DESCRIPTION
**Pull request by xclaesse**
When a library 'foo' has a public dependency 'bar', foo.pc will be
generated with something like this:
bar_dep = dependency('bar', fallback : ['bar', 'bar_dep'])
libfoo = library('foo', dependencies : bar_dep)
pkg.generate(libfoo, libraries : bar_dep)
    
In the subproject we will typically have this (not using pkg.generate):
libbar = library('bar')
bar_dep = declare_dependency(link_with : libbar)
configure_file(input : 'bar.pc.in', output : 'bar.pc', ...)
    
When the 'bar' dependency is built as subproject, libraries from
InternalDependency were not processed. If we process them,
then libbar.generated_pc is suddenly set to 'foo', which will cause any other
pc file generated to have 'Required : foo' for the 'bar' dependency. To
fix that 2nd issue, we set generated_pc only on the main lib passed a
positional argument, because in that case the intention is obvious. Libs
passed to "libraries" could be just public dependencies and not the
libraries for which we generate a pc file.

### CODE REVIEW

**Comment by xclaesse** [comment]
@jpakkane @nirbheek Would be great to have a review, I have patches blocked by this for GStreamer.

**Comment by xclaesse** [comment]
@jpakkane @nirbheek ping?

**Comment by jpakkane** [comment]
Since we are touching this part, could we check and never add a pc file entry to a library that is built but which does not get installed?

We should probably update the pkgconfig module documentation to describe what dependencies the system decides to add and how.

**Comment by xclaesse** [comment]
> Since we are touching this part, could we check and never add a pc file entry to a library that is built but which does not get installed?

I think we should merge https://github.com/mesonbuild/meson/pull/3939 first, that already fix the most common case where internal static libraries are being added into the pc file. Once that's fixed, we could maybe make it hard error to have uninstalled libraries in the pc generator?

> We should probably update the pkgconfig module documentation to describe what dependencies the system decides to add and how.

Yes, definitely.

**Comment by xclaesse** [comment]
Note: I sneaked `BuildTarget: Fix typo, self.install does not exists` commit unrelated to this PR, but found that typo while working on this.

**Comment by xclaesse** [comment]
@jpakkane I tried to document the rules as much as I can, but it's difficult. Many subtle cases.

File: docs/markdown/Pkgconfig-module.md [hunk 1]
```diff
@@ -59,3 +61,38 @@ provided for all required fields of the pc file:
...
+
+- Not found libraries or dependencies are ignored.
+- Libraries and dependencies are private by default (i.e. added into
+  `Requires.private:` or `Libs.private:`) unless they are explicitly added in
+  `libraries` or `requires` keyword arguments, or is the main library (first
+  positional argument).
+- Libraries and dependencies will be de-duplicated, if they are added in both
+  public and private (e.g `Requires:` and `Requires.private:`) it will be removed
+  from the private list.
+- Shared libraries (i.e. `shared_library()` and **NOT** `library()`) add only
```
Comment (textshell) [inline docs/markdown/Pkgconfig-module.md]: I don't remember if this is true for `library()` when building with default shared libraries. We should check this.

Comment (textshell) [reply docs/markdown/Pkgconfig-module.md]: Thanks for reminding me. 

I *think* the rational was that we want the .pc files to be the same regardless of default library type.

Comment (xclaesse) [reply docs/markdown/Pkgconfig-module.md]: Yes that's the rational IIRC. It makes sense, even if it's not obvious at all, so it's good that it finally gets documented at least.

File: mesonbuild/modules/pkgconfig.py [hunk 2]
```diff
@@ -114,14 +122,8 @@ def _process_libs(self, libs, public):
                 # than needed build deps.
                 # See https://bugs.freedesktop.org/show_bug.cgi?id=105572
                 processed_libs.append(obj)
-                if public:
-                    if not hasattr(obj, 'generated_pc'):
-                        obj.generated_pc = self.name
```
Comment (textshell) [inline mesonbuild/modules/pkgconfig.py]: I think this is a breaking change. I believe it's a change in the right direction, but we should at least have a release note snippet that tell the users how to change their build definitions.

I'm not sure if there is a graceful way to deprecate this and only change over at a later point.

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: Yes that's an intentional break, and indeed that must be added into release notes.

**Comment by textshell** [comment]
> Since we are touching this part, could we check and never add a pc file entry to a library that is built but which does not get installed?

Let's keep iterating this in small steps and react to feedback and breakages one piece at a time. I think this PR is big enough as is.

File: mesonbuild/modules/pkgconfig.py (hunk 2 above)
Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: Note, since it wasn't documented previously, and it's pretty corner case, I think it's not that bad.

Comment (xclaesse) [reply mesonbuild/modules/pkgconfig.py]: Release notes added.

**Comment by textshell** [comment]
This definitively breaking, but i think removing the implicit "associate every internal library seen with the generated pkg-config file" needs to be more explicit.

> Note, since it wasn't documented previously, and it's pretty corner case, I think it's not that bad.
It's for all uses that did not yet migrate to using a positional for the main library.

I wonder if we can have a better deprecation story here though. If should definitively have the part that associates the main library. Maybe we could for a transition period associate also the directly given public dependencies while issuing a warning that this is deprecated.

What about this?

```
        if mainlib:
￼             deps.add_pub_libs(mainlib)
￼             if not hasattr(mainlib, 'generated_pc'):
￼                 mainlib.generated_pc = filebase

        # Remove this loop after deprecation period
        for direct_dep in mesonlib.listify(kwargs.get('libraries', []), unholder=True):
            if isinstance(direct_dep, (build.SharedLibrary, build.StaticLibrary)):
                if not hasattr(obj, 'generated_pc'):
                        mlog.deprecation('library passed via libraries: argument implicitly associated with pkg-config file. Pass the main library as positional argument')
                        obj.generated_pc = self.name
```

File: docs/markdown/snippets/pkgconfig_break.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+## Behaviour change in pkg-config generator
+
+A small behaviour change has been made in the pkg-config generator. It will only
+generate a `Requires:` or `Requires.private:` for a `library()` if is it passed
+as first positional argument of the `generate()` method, and not any more if it
+is passed to the `libraries` keyword argument any more.
```
Comment (textshell) [inline docs/markdown/snippets/pkgconfig_break.md]: I think this paragraph is hard to understand correctly.

We don't seem to have any wording at all about this important effect of associating internal libraries with the generated pkg-config files.

I think i would rephrase this as using the positional argument for the main library is now mandatory. That is very actionable for our users and we can then after this simple case add a caveat about previously allowing multiple libraries to be associated with one pkg-config file and that this will no longer be possible. (If we go with the depreciation idea i proposed it will still be possible for a time but we should urge users to get in contact with their use case anyway to see if we are missing something)

**Comment by xclaesse** [comment]
@textshell it's tricky because I would like to avoid warnings in valid use cases. For example your suggestion would warn on this code:
```meson
liba = library()
libb = library(link_with : liba)
pkg.generate(libb, libraries : liba)
```

File: docs/markdown/snippets/pkgconfig_break.md (hunk 3 above)
Comment (textshell) [reply docs/markdown/snippets/pkgconfig_break.md]: Yes, of course. It's only mandatory for pkg-config files that represent built libraries. But i still think building around that would be more understandable. But if you come up with something else that's easy to understand that's fine too. But i don't like how the current one focuses on generated lines and not on something high level. Also i had a hard time understanding it and i worked on this code. Maybe it could also phrased something like "to allow further dependencies on a library for which a pkg-config file was generated with pkg.generate to refer to the generated file..."

Comment (xclaesse) [reply docs/markdown/snippets/pkgconfig_break.md]: I tried to rephrase it, does it look better? I'm not really good at writing doc, tbh :(

**Comment by xclaesse** [comment]
I updated this PR to also fix #4583 by setting `generated_pc` only after generating the current pc file, to make sure it won't refers to itself.

> > @textshell it's tricky because I would like to avoid warnings in valid use cases. For example your suggestion would warn on this code:
> > ```meson
> > liba = library()
> > libb = library(link_with : liba)
> > pkg.generate(libb, libraries : liba)
> > ```
> 
> Well in that case it will change behavior if liba is used later. Of course we could improve the precision by not warning there but adding a attribute `generated_pc_warning` with the location and warning text and changing `_process_reqs` to display the warning in the `hasattr(obj, 'generated_pc')` block.

(oh, github has "quote reply", is that new?)

The point here is that you could write that snipped in new code intentionally, and that would warn. I've been thinking about this, and I see no way we could add warnings and still avoid printing warnings if the usage is intentional. But I have to admit I'm a bit confused myself, so maybe you've got better idea?

File: docs/markdown/Pkgconfig-module.md [hunk 4]
```diff
@@ -25,14 +28,13 @@ keyword arguments.
...
-  `Requires.private` instead. Other type of dependency objects can also be passed
-  and will result in their `link_args` and `compile_args` to be added to `Libs`
-  and `Cflags` fields.
+  dependencies of built libraries will be automatically added, see the
+  [Implicit dependencies](#Implicit_dependencies) section below for the exact
+  rules.
 - `libraries_private` list of built libraries or strings to put in the
-  `Libs.private` field. Since 0.45.0 it can also contain dependency objects,
-  their `link_args` will be added to `Libs.private`.
+  `Libs.private` field. Since 0.45.0 dependencies of built libraries will be
```
Comment (bluca) [inline docs/markdown/Pkgconfig-module.md]: I think this actually happened in 0.46 - if that's the case, what about fixing it as well while this line is being changed?

https://github.com/mesonbuild/meson/blob/master/docs/markdown/Release-notes-for-0.46.0.md#improved-generation-of-pkg-config-files-for-static-only-libraries

File: docs/markdown/snippets/pkgconfig_break.md (hunk 3 above)
Comment (xclaesse) [reply docs/markdown/snippets/pkgconfig_break.md]: Changed this text to be a deprecation warning instead of behaviour change.

**Comment by xclaesse** [comment]
@textshell Ok, I changed to keep previous behaviour and print deprecation warning. I tested a few projects I'm contributing to, and the deprecation warning happens, but is fixable in every cases. So I think we are good.

File: docs/markdown/snippets/pkgconfig_break.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+## Deprecation warning in pkg-config generator
+
+All libraries passed to the `libraries` keyword argument of the `generate()`
+method used to be associated with that generated pkg-config file. That means
+that any subsequent call to `generate()` where those libraries appear would add
+them into `Requires:` or `Requires.private:` field instead of `Libs:` or
```
Comment (textshell) [inline docs/markdown/snippets/pkgconfig_break.md]:
```suggestion
the filebase of the `generate()` that first contained them into `Requires:` or `Requires.private:` field instead of adding an `-l` to `Libs:` or
```

File: docs/markdown/snippets/pkgconfig_break.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+`Libs.private:`.
+
+This behaviour is now deprecated. The library that should be associated with
+the generated pkg-config file should be passed as first positional argument
+instead of in the `libraries` keyword argument. The previous behaviour is
+maintained but prints a deprecation warning and could change in a future Meson
```
Comment (textshell) [inline docs/markdown/snippets/pkgconfig_break.md]:
```suggestion
maintained but prints a deprecation warning and support for this will be removed
in a future Meson
```

**Comment by jpakkane** [comment]
With this are there any known problems with pgkconfig generation remaining? That is, do we need to do major changes in the future (that we know of) or will they be minor tweaks?

File: docs/markdown/snippets/pkgconfig_break.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+```
+
+It can be fixed by passing `liba` as first positional argument::
+```meson
+liba = library(...)
+pkg.generate(liba)
+
+libb = library(...)
+pkg.generate(libb, libraries : [liba])
+```
```
Comment (textshell) [inline docs/markdown/snippets/pkgconfig_break.md]: +

"""If you can not create the needed pkg-config file without this warning, please file an issue with as much details as possible about the situation"""

**Comment by textshell** [comment]
I'll try to summarize this change to make sure we are all on the same page here.

We have 3 major changes:
* New feature: Linker and compiler parameters supplied with declare_dependency are included in the generated file.
* Deprecation: Passing the primary library that should be associated with an generated pkg-config file as a direct public dependency instead of as the (only) positional parameter is now deprecated and will cause an warning. This means all projects that did not yet switch to this feature introduced in 0.46 *and* that use such an library as dependency in a further generate need to be adjusted to avoid the deprecation warning.
* Removal: Before this change public dependencies where recursively considered to be associated with the generated pkg-config file. While associating direct dependencies are covered but the deprecation above support for implicitly associating indirect dependencies is removed in this PR. Depending on that should be rare and was not considered to be important enough to complicate the deprecation code to also cover.

Also @xclaesse improved the documentation a lot.

**Comment by textshell** [comment]
> With this are there any known problems with pgkconfig generation remaining? That is, do we need to do major changes in the future (that we know of) or will they be minor tweaks?

I'm not currently aware of anything. But we are doing build systems, so always expect playing whack-a-mole. But lets try not to bog down this PR which speculation about future steps that are only mildly related. Because as the saying goes: "How to make a great build system? – By solving one edge case at a time."

**Comment by jpakkane** [comment]
LGTM. Ok to merge even though the freeze DL has passed. It would be great if we could get a quick review from someone with experience with pkg-config (I don't have much knowledge about advanced pgk-config usage scenarios). But if we don't get one soon, go ahead and merge.

**Comment by textshell** [comment]
From my perspective the only open points are minor improvements in the wording of the release notes and documentation. Apart from that this is reviewed by me and looks good. Of course additional review is always good.

**Comment by xclaesse** [comment]
>     * New feature: Linker and compiler parameters supplied with declare_dependency are included in the generated file.

Actually linker and compiler parameters were already pulled, because InternalDependency is a subclass of Dependency. Here we are adding libraries, those from declare_dependency(link_with:foo).

>     * Removal: Before this change public dependencies where recursively considered to be associated with the generated pkg-config file. While associating direct dependencies are covered but the deprecation above support for implicitly associating indirect dependencies is removed in this PR. Depending on that should be rare and was not considered to be important enough to complicate the deprecation code to also cover.

I don't think that changed actually. Recursive dependencies are added as private and those where not associated with the generated pkg-config file. All public libraries are still associated. AFAIK.

> With this are there any known problems with pgkconfig generation remaining? That is, do we need to do major changes in the future (that we know of) or will they be minor tweaks?

There is that story about not installed libraries that sometimes get included, but that should be done in a separate PR and will be mitigated by https://github.com/mesonbuild/meson/pull/3939 if @nirbheek finds the time to finish it.

Other than that, I think we are good... until someone reports a bug...

File: docs/markdown/snippets/pkgconfig_break.md (hunk 7 above)
Comment (xclaesse) [reply docs/markdown/snippets/pkgconfig_break.md]: done.

**Comment by xclaesse** [comment]
@jpakkane @textshell I think all comments have been addressed now. Looks good to me.
